### PR TITLE
[runtime env] Download runtime env(conda) in agent instead of setup_worker

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?

- Download runtime env(conda) in agent instead of setup_worker
- Add `--runtime-env-setup-hook` argument to `ray start` script to make it pluggable.
- Add a retry mechanism in agent.

### CODE REVIEW

**Comment by edoakes** [comment]
Thanks for following up on this @SongGuyang I'm really excited to have this cleaned up and the behavior centralized in the agent :)

One thing I'm not sure about the status of right now -- what happens when the agent fails to create/install and environment? Ideally this would cause the submitted task to be failed with a relevant error message. Prior to the agent change this basically hangs forever and it has been on our backlog, not sure if you're already handling this in the agent.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (architkulkarni) [inline dashboard/modules/runtime_env/runtime_env_agent.py:40]: It looks like the runtime env dict is being used as a way to communicate from the agent setup function to the worker_setup function, probably because the existing implementation of the setup function was unfortunately not very pluggable.  It seems fine for now, but do you have any ideas on what the long term solution should be for this kind of communication? Or is this okay for the long term?  I'm just wondering if we need some kind of better API for this.

**Review by architkulkarni — COMMENTED** [review]
Thanks for this PR, it looks like a really nice refactor!  I left a few comments and questions.  I think it would be valuable to get @iycheng's thoughts as well, since he has some experience working with the shim process and might have some ideas on how to improve the design.

**Comment by SongGuyang** [comment]
@edoakes @architkulkarni About failure of runtime env creating in agent, we don't have a complete feature currently. I think we can do like this:
- Add retry mechanism in agent internal. This will resolve some short-time faults when create runtime env, such as package download timeout. I can add this retry in current PR.
- If also fails after some retry times, return an error code and error message to raylet. Raylet should process this error and reply an error to gcs server by `RequestWorkerLeaseReply`. And gcs server should catch this error and rescheduling to other nodes(avoid to reschedule to the same node). I can not implement this issue in current PR because we should make a refactor of `PopWorker` in worker pool. We need a async `PopWorker`, but it is a sync function now. We plan to make another PR to do this.
- Another mechanism we should add is that make a timeout in raylet when wait for runtime env creation reply. If some serious faults happened in agent, such as unavailability of network, raylet should also trigger gcs server to reschedule. I think this issue also will be implemented in feature PR.

**Comment by SongGuyang** [comment]
@edoakes @architkulkarni And I found another problem. We should make `setup_runtime_env` to be an async function. Otherwise, it will block agent. I should take a look how to do this.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:899]: I don't add hash because this is a dedicated worker which will be looked up by task id. But I also can add this hash in current PR.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:46]: Yep, I also have thought about this question. But I didn't find any better way to do this because  `setup_worker` just parses params from command line.

**Comment by edoakes** [comment]
@SongGuyang the failure handling logic sounds pretty good to me at a high level except that eventually we should fail the task if there's an issue in setting up the environment. FYI we have an issue tracking this feature here:
https://github.com/ray-project/ray/pull/16525

File: python/ray/workers/setup_runtime_env.py
Comment (SongGuyang) [reply python/ray/workers/setup_runtime_env.py:90]: Ok, I have fixed this.

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [reply src/ray/raylet/worker_pool.cc]: Ah, you're right--thanks for the clarification, I forgot this was the code path for dedicated workers.

**Comment by architkulkarni** [comment]
I just remembered the shim process is used to set up the environment for the Ray Client server--does this part also need to be modified to work with this PR, or is it okay?  https://github.com/ray-project/ray/blob/master/python/ray/_private/services.py#L1947-L1962

**Comment by SongGuyang** [comment]
> I just remembered the shim process is used to set up the environment for the Ray Client server--does this part also need to be modified to work with this PR, or is it okay? https://github.com/ray-project/ray/blob/master/python/ray/_private/services.py#L1947-L1962

@architkulkarni Oh, I haven't paid attention to this. Maybe this is the reason why a lot test failed. Let me see and modify this. Thank you so mach!

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.cc:834]: GetRuntimEnvHash -> GetRuntimeEnvHash

**Comment by architkulkarni** [comment]
I think the only possibly relevant test failure that isn't known to be flaky is `test_runtime_env_env_vars`.  https://buildkite.com/ray-project/ray-builders-pr/builds/7647#5ab5cb9f-b0f6-40c5-a47a-10b240b8d328/217-528. What's odd is that the test that's timing out is `test_override_environment_variables_task[False]`, where `False` means it isn't using runtime_env at all, but is instead using the legacy `override_environment_variables` feature.  I'm about to make a PR to remove the `override_environment_variables` feature anyway, but is it possible there's some issue involving the hash of the WorkerCacheKey?

**Comment by SongGuyang** [comment]
@architkulkarni I have fixed `test_override_environment_variables_task[False]`. It is my fault which loses `runtime_env_hash`.  https://github.com/ray-project/ray/pull/16525/commits/1a084a39e9049356c86fe560a5a35ddad0a03166#diff-4c19d6ebd8575189e9439fa5646b567546dc54f64d576a959ce1bb43902d1a03R964

**Comment by architkulkarni** [comment]
Only failing tests are linux://python/ray/tests:test_scheduling and windows://python/ray/tests:test_array which are known flaky according to https://flaky-tests.ray.io/.  There's a merge conflict unfortunately, but after resolving that I think this is ready to merge!

File: python/ray/workers/setup_runtime_env.py
Comment (raulchen) [inline python/ray/workers/setup_runtime_env.py:93]: would the result be empty? If yes, print a warning or something? If not, remove this if

**Review by raulchen — APPROVED** [review]

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (raulchen) [reply dashboard/modules/runtime_env/runtime_env_agent.py:69]: I think we should separate the "runtime env spec" (which is specified by the user) and the "runtime env context (which contains the context info about a runtime env, such as path to a conda). 
Instead of inserting the context info back the the `RuntimeEnv` object, it's better to define a new class `RuntimeEnvContext` and use a new field in the reply. 
This would help improve code readability and maintainability.

**Comment by raulchen** [comment]
(This can be done in the next PR.) Another suggestion about the plugin system. 
Currently, we make runtime env pluggable by exposing some hook APIs. The main problem is that if users want to extend a new runtime env type, they will have to replace the whole hook. So it's hard for them to use both the Ray built-in types and their custom types simultaneously. 

So I'd propose changing the plug-in system in the way that one plugin only handles one runtime env type. E.g., we define a `RuntimeEnvPlugin` interface. And users can implement their plugins with sub-classes. 

The interface will be something like this.
```
class RuntimeEnvPlugin:
      @abstractmethod
      def runtime_env_type(self):
            pass
            
       @abstractmethod
       def setup_runtime_env(...):
            pass
            
        @abstractmethod
       def start_worker(...):
             pass
````

Let's say, a user implements a new plugin that can handle type X. Then they can use this type in code `ray.init(runtime_env={"type": "X"}). At runtime, ray will use the X-type plugin to set up the runtime env and start the workers.

cc @edoakes @architkulkarni

**Comment by architkulkarni** [comment]
Thanks for the suggestion, that sounds like a good idea.  I agree that the current pluggability is hard to use for the reason you mentioned (we have an issue for it here https://github.com/ray-project/ray/issues/16035)

Just to clarify, you're saying that to use the custom type together with the built-in types, the implementation would do some custom setup and then call `super.setup_runtime_env()` to do the setup for the Ray built-in types, right?  I'm also wondering if it's worth trying to support arbitrary combinations of custom types instead of just one custom type at a time, but maybe that's too complicated and there isn't a real need for it.

**Review by architkulkarni — APPROVED** [review]

**Comment by raulchen** [comment]
> Just to clarify, you're saying that to use the custom type together with the built-in types, the implementation would do some custom setup and then call super.setup_runtime_env() to do the setup for the Ray built-in types, right?

No, the super class only defines the framework, e..g, inputs and outputs of the plugins. It doesn't have any concrete code, so we don't need to call `super.setup_runtime_env()`. 

To make it more clear, let's say Ray already supports Conda with `CondaRuntimeEnvPlugin`. If a user wants to use VirtualEnv,  they can implement a new plugin called `VirtualEnvRuntimeEnvPlugin`. Then in the same cluster, some jobs can use conda with `runtme_env={"type": "conda"}`, and some jobs can use virtualenv with `runtime_env={"type": "virtualenv"}`. At runtime, when starting a worker, Ray will check the runtime env type first. If the type is "virtualenv", Ray will use `VirtualEnvRuntimeEnvPlugin.setup_runtime_env` to set up the runtime env for this worker.

All plugins are orthogonal, so we don't need to deal with "combinations" of these plugins.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:33]: `self._setup` has been used in this function.

**Pull request merged**
